Course hole distances

vincecampanale opened the pull request:
# What / why

- commits new object `courseHoleDistances` which holds all the distances of each hole on each course from each tee to each pin, used in https://github.com/Golf-Scope/golf-plus-api/pull/1308
- adds script that can be used to generate this object from the raw course json objects which we also keep in `vr-web-ui`
- adds a description of the script, how to use it to generate new `courseHoleDistances`, and when we should run it

vincecampanale commented on scripts/data/raw-course-json/alpine.json:
```diff
@@ -0,0 +1,270 @@
+{
+  "id": "alpine",
```
`id` seems to always match the value we put in our `COURSES` constant but I'm not sure why so would like to ask where this `id` comes from. Who creates and uploads these json files?

vincecampanale commented on scripts/data/raw-course-json/alpine.json:
```diff
@@ -0,0 +1,270 @@
+{
```
These json files are duplicated from vr-web-ui. If we change one for some reason - V2 of a course perhaps? - then we have to replace the file here. Open to better ways to manage this to maybe require less manual steps. Although it isn't that much to put the file in here as part of running the script.

vincecampanale commented on README.md:
```diff
@@ -1,3 +1,11 @@
+## get-hole-distances-from-course-json.js
+* Processes raw course JSON files into course hole distances. Any time we need to know the distance from tee to pin for a given hole, we can simply reference this object. 
+* When we add a new course json file to vr-web-ui or when we change an existing one, we also need to copy it into the `scripts/data/raw-course-json` directory in this repo and run this script.
+* The script will write the new course hole distances constant into `src/course-hole-distances.js` but it will be unformatted. Be sure to format the file after running the script so the diff is clear. One easy option for formatting is to go into the file and add a new line after `module.exports = {` then save with Prettier "format on save" enabled.
```
The high level process would be:

- drop the new json file in `data/raw-course-json` 
- `node get-hole-distances-from-course-json.js`
- format `src/course-hole-distances.js` and eyeball the diff 

I tested this by changing some things locally and it worked. I think the `.prettierrc` in here should result in us both producing the same diffs.

seankarson replied:
Why doesn't our precommit hook take care of the formatting?

vincecampanale replied:
It looks the same as the pre commit hook we have set up in golf-plus-api...not sure what's up with it. I can add this as a follow up in my column

vincecampanale commented on src/course-hole-distances.js:
```diff
@@ -0,0 +1,6163 @@
+  alpine: {
+    1: {
+      front: {
+        easy: 230.153533935547,
```
Should we round these since we're gonna be using them in the API? Don't see a need for all the decimals

seankarson replied:
I don't think it matters much, but I'd prefer rounding later on since we might be summing some of these up

vincecampanale commented on scripts/data/course-json/alpine.json:
```diff
@@ -0,0 +1,270 @@
+{
+  "id": "alpine",
```
id seems to always match the value we put in our COURSES constant but I'm not sure why so would like to ask where this id comes from. Who creates and uploads these json files?

seankarson replied:
fairly certain the course team does this and then notifies us of the course id

vincecampanale commented on scripts/data/course-json/alpine.json:
```diff
@@ -0,0 +1,270 @@
+{
```
These files are duplicated from vr-web-ui. If we change one - was valhalla V2 an example of this? - then we have to replace the file here. Open to better ways to manage this to maybe require less manual steps. Although it isn't that much to put the file in here as part of running the script.

seankarson commented on scripts/data/course-json/clubhouse.json:
```diff
@@ -0,0 +1,16 @@
+{
```
don't need this one :)

seankarson commented on scripts/get-hole-distances-from-course-json.js:
```diff
@@ -0,0 +1,54 @@
+    });
+  }
+
+  const data = new Uint8Array(
```
why can't you just write the string directly? do you need to use this buffer approach?

vincecampanale replied:
just copied it from node docs, dropped it and the script works fine

seankarson commented on scripts/get-hole-distances-from-course-json.js:
```diff
@@ -0,0 +1,54 @@
+  }
+
+  const data = new Uint8Array(
+    Buffer.from(`module.exports = ${JSON.stringify(courseHoleDistances)}`)
```
JSON.stringify takes pretty print parameters if you want to add those

vincecampanale replied:
added, doesn't remove the need to run prettier on the file though - looking into why our pre commit hook doesn't format for us

seankarson commented:
This branch has butler, muirfield, and st. andrews reverse - let's add those

https://github.com/Golf-Scope/vr-web-ui/tree/post-dt-cleanup/src/courses/data

Can you also add a note to the Adding a New Course notion doc to mention checking for last minute changes to the json files at course release time (in case distances get updated)

seankarson reviewed (commented):
What you have is totally fine - just asked for a few minor changes

vincecampanale commented:
@seankarson back to you

- changed the object shape per your suggestion, I agree it is cleaner
- exported a function to get the data instead of the object itself
- added the new courses
- added pretty printing params but the json output still needs to be formatted, mainly just to drop the quotes
- not sure why the pre commit hook isn't doing the formatting

vincecampanale commented:
Also, updated the "Adding a new course" Notion doc: [here](https://www.notion.so/golfplus/Adding-a-new-course-94f10d0816d143fc9e5692ac78e7c635?pvs=4#6346c50101a84bda8f72245e85742282) and [here](https://www.notion.so/golfplus/Adding-a-new-course-94f10d0816d143fc9e5692ac78e7c635?pvs=4#263ca95b52d04c9d921937d2d1e98daf)

seankarson commented on src/index.js:
```diff
@@ -11,6 +11,27 @@ const {
+    throw new Error('Hole must be integer between 1 and 18');
+  }
+
+  return courseHoleDistances[course][hole - 1][`${tee}_${pin}`];
```
let's bomb if this doesn't exist - thinking of a call to get hole 10 data for butler, for instance

vincecampanale replied:
I think a call to get a non existent hole would bomb with a "cannot get tee_pin of undefined" type error but I added an error if holeDistance == null. Does that work for your suggestion or are you looking for more validation

seankarson reviewed (approved):
just one comment + the husky stuff we discussed in slack, then good to go
